Fix Inconsistent droid requirements in an all AI game

Mahoyomu opened the pull request:
### Goal of the PR
- Fix #9879
### How does the PR work?
- By adding a `force` action and related code to command `droid`.
### Does it resolve any reported issue?
- #9879
### If not a bug fix, why is this PR needed? What use cases does it solve?
- This is a bug fix.

## To do

Adjust possibly missed related method and changes I mistakenly made, remove redundant notes I used for quickly locating(all after this pr is nearly ready for review).

This PR is Ready for Review.

## How to test
Compile and run, see the test video I might upload later if needed, or see changes in the src.

ProditorMagnus commented:
If changing control is technically allowed, there is no need to make user just type more for it. I got impression reason for the error was sync issue.

Mahoyomu commented:
> If changing control is technically allowed, there is no need to make user just type more for it. I got impression reason for the error was sync issue.

Alright. I can also implement it a way that won't use this force param

Mahoyomu commented:
> If changing control is technically allowed, there is no need to make user just type more for it. I got impression reason for the error was sync issue.

Btw, is it fine if I make it possible for user to droid any side any time? I'm actually a bit confused by the contradiction of `is_your_turn` in the related codes and requirement in the issue.

According to the commit message provided in https://github.com/wesnoth/wesnoth/commit/204357c6d2b722bf5765002a52da8a98bad1a4e0
, it seems `droiding a side` should only work when `you own` it. My fix is based on that message, it prevents droiding a side players actually don't own(fully controlled by AI), including observers, whereas they are actually considered to own side 1 before the fix. But they can still explicitly do it by adding a force param.

Could you please provide me with more details about the direction I should go to? Also fine for me to allow droiding any side tho, just asking(because I'm a little confusing)

ProditorMagnus commented:
My understanding of AI is that all sides that are controlled by AI which runs in your wesnoth process, should be considered owned by you.

Mahoyomu commented:
> My understanding of AI is that all sides that are controlled by AI which runs in your wesnoth process, should be considered owned by you.

That means if some player execute `droid full`, they are also ai right now, and they will also be considered owned by any player?

ProditorMagnus commented:
There is no centralized AI. One of the involved clients has to own AI. Unless you change something, they are owned by host.

Mahoyomu commented:
> There is no centralized AI. One of the involved clients has to own AI. Unless you change something, sides that started as AI are owned by host. Host is the only one that can droid them.
> 
> Sides that later become AI are owned by the client that last controlled them. That client is the only one that can droid them. If host wants to do that, he first needs to take control of the side.
> 
> I am not aware of droid full.

Appreciate for your explaination ;)

Mahoyomu commented:
Test video:

https://github.com/user-attachments/assets/7c2e8721-f0a3-47b6-bf75-33e96fb2edbe

It seems to work in this setup now.

soliton- commented on src/play_controller.hpp:
```diff
@@ -168,6 +168,11 @@ class play_controller : public controller_base, public events::observer, public
 		return gamestate().board_.is_observer();
 	}
 
+	bool is_singleplayer() const
```
Is there a reason for this method when `is_networked_mp()` exists?

Mahoyomu replied:
I missed it. Sorry

soliton- commented on src/menu_events.cpp:
```diff
@@ -1411,15 +1411,25 @@ void console_handler::do_refresh()
 	// default to the current side if empty
 	const unsigned int side = side_s.empty() ? team_num_ : lexical_cast_default<unsigned int>(side_s);
-	const bool is_your_turn = menu_handler_.pc_.current_side() == static_cast<int>(menu_handler_.gui_->viewing_team().side());
+	const bool is_your_turn = menu_handler_.pc_.is_singleplayer() ? true : menu_handler_.pc_.current_side() == static_cast<int>(menu_handler_.gui_->viewing_team().side());
```
I don't think it's readable to set `is_your_turn` to true when it is not your turn.

soliton- commented:
IMO honestly best would be to revert 204357c. It added complexity for questionable gain. `[side_filter]` for example ignores `controller` still.

Pentarctagon commented:
The bug it fixed, #4996, was marked as a new stable release blocker though.

soliton- commented:
It was also marked as a regression when it wasn't one.

soliton- commented on src/menu_events.cpp:
```diff
@@ -1411,15 +1411,25 @@ void console_handler::do_refresh()
 {
-	// :droid [<side> [on/off/full]]
-	const std::string side_s = get_arg(1);
+	// :droid [side] [on/off/full]
```
`<side>` means that you have to replace this with an appropriate value. In contrast to `on/off/full` (which should really be `on|off|full]` as in the help) where those are the literal values to use.

If you must change how the command works then please do it in a separate commit.

Pentarctagon commented:
I don't think it's a good idea to change it back to being unsafe to use - rather than making this even more complicated, it should just not allow droiding.

Mahoyomu commented:
> I don't think it's a good idea to change it back to being unsafe to use - rather than making this even more complicated, it should just not allow droiding.

By saying `not allow droiding`, you mean when using the setup mentioned in #9879 , just don't allow any droiding command?

Mahoyomu commented:
Should I do a regression or not? I'm a little confused by the difference of opinion now.

soliton- commented:
> I don't think it's a good idea to change it back to being unsafe to use - rather than making this even more complicated, it should just not allow droiding.

That's a lot of its.

My guess is that people read about WML that might cause OOS and think it can simply be fixed to be "safe" to use and that's that. Which is simply not the reality. There is no way around understanding in detail what causes OOS and what doesn't. Which also includes understanding that OOS is not MP-exclusive.

The only way to not have to understand anything is to simply forbid everything that in any way could be misused. Which in essence probably means to remove replay and multiplayer functionality.

As far as I can tell 204357c made it "safe" to specifically check for controller `ai` after `:droid`. It remains "unsafe" in all other cases. It also made it so the check for controller `ai` (after `:droid`) can now not be used to figure out a side that is actually controlled by the AI on the current client (which IMO is the most sensible use of this check).

Also seems weird that clients that do not own a side now in some capacity think it's an AI side. This is different from a side being AI from the start, right? This all seems more confusing than useful.

Pentarctagon commented:
Well, however you want to solve it then. I don't really have a strong opinion either way.

Mahoyomu commented:
Got it. If that's the case I'll start doing regression soon. Appreciated for ideas shared ;)

Mahoyomu commented:
> > I don't think it's a good idea to change it back to being unsafe to use - rather than making this even more complicated, it should just not allow droiding.
> 
> That's a lot of its.
> 
> My guess is that people read about WML that might cause OOS and think it can simply be fixed to be "safe" to use and that's that. Which is simply not the reality. There is no way around understanding in detail what causes OOS and what doesn't. Which also includes understanding that OOS is not MP-exclusive.
> 
> The only way to not have to understand anything is to simply forbid everything that in any way could be misused. Which in essence probably means to remove replay and multiplayer functionality.
> 
> As far as I can tell [204357c](https://github.com/wesnoth/wesnoth/commit/204357c6d2b722bf5765002a52da8a98bad1a4e0) made it "safe" to specifically check for controller `ai` after `:droid`. It remains "unsafe" in all other cases. It also made it so the check for controller `ai` (after `:droid`) can now not be used to figure out a side that is actually controlled by the AI on the current client (which IMO is the most sensible use of this check).
> 
> Also seems weird that clients that do not own a side now in some capacity think it's an AI side. This is different from a side being AI from the start, right? This all seems more confusing than useful.

I've tested it and proved what you've said, and after I went through related issues and PRs, I probably got what you mean. The check for controller `ai` must be combined with the check for `networked` to figure out if a side is controlled locally by AI (and the check for sides initially controlled by AI also doesn't work now because `:droid full` can change that).

Maybe it's better to not touch the controller attribute (meaning don't allow manually changing the controller attribute from human to AI by `:droid` or vice versa), after all, by the name of `:droid`, it should just use AI to help players act, and that's also the only thing it should do, not including changing the controller status.

By the way, in the regression commit you mentioned, you said you wanted an overhaul of `:droid`. Could you please give me more details about the overhaul you have in mind? I also noticed that Celmin once said `I still think an added option in :give_control would've been preferable...` in #4147, which introduced OOS problems after adding `full` to `:droid`. I'm wondering if this can be implemented in the way Celmin proposed and also avoid the OOS problems that this `full` option introduced.

Mahoyomu commented:
Actually.... When I was testing if the fix for #3665 works, I found out it seems I can't see the sides I originally control. I can't see them until it's my turn again. I'm not sure if the added `full` really solved that issue too, and maybe I just missed something or tested in an improper way....

Mahoyomu commented:
> > Also seems weird that clients that do not own a side now in some capacity think it's an AI side. This is different from a side being AI from the start, right? This all seems more confusing than useful.
> 
> It's not different from the start. Ai sides controllted by another client at start appear as "ai" controller.
> 
> There are still cases where where checking controller can cause OOS thought, for example the tracking of controller changes in [204357c](https://github.com/wesnoth/wesnoth/commit/204357c6d2b722bf5765002a52da8a98bad1a4e0) is not stored in the replay, also it is still possible that undoable actions can be executed before the controller changed at one client and after it on another client. Furthermore it can probably also cause OOS if :droid is used and then a player uses the "replay turn" feature.
> 
> also most of these issues probaly also apply to the "replace by ai" option when a player leaves.
> 
> > after all, by the name of :droid, it should just use AI to help players act, and that's also the only thing it should do, not including changing the controller status.
> 
> That's what we originally thought when we changed the bahaviour (around 1.14 i think?). But players then really wanted a way to give conrtol away from them thats why the "full" option was added.

And they can use AI option after executing ":give_control", like what celmin said.(if we implemented this way)

ProditorMagnus commented:
For WML use about controller and OOS, my current understanding is that controller key should not be used, and instead `wesnoth.sync.evaluate_single` is the safe way to ask correct client, does it consider itself ai.

soliton- commented:
> > Also seems weird that clients that do not own a side now in some capacity think it's an AI side. This is different from a side being AI from the start, right? This all seems more confusing than useful.
> 
> It's not different from the start. Ai sides controllted by another client at start appear as "ai" controller.

Oh, I expected that to be the actual controller of the side. In a test even human sides are the same on all clients? Was it always this way? Makes it really confusing to me why a side filter for example completely ignores a controller attribute in networked play and you can find warnings on the wiki about side controller being different in multiplayer.

Really seems like this was changed at some point and no other code or documentation updated...

> > after all, by the name of :droid, it should just use AI to help players act, and that's also the only thing it should do, not including changing the controller status.
> 
> That's what we originally thought when we changed the bahaviour (around 1.14 i think?). But players then really wanted a way to give conrtol away from them thats why the "full" option was added.

Really originally "full" was the only thing. At some point a proxy controller was introduced and then you could not change control fully at all until the introduction of the "full" parameter. I'm not sure all this complication was a step in the right direction...

Mahoyomu commented:
> Really originally "full" was the only thing. At some point a proxy controller was introduced and then you could not change control fully at all until the introduction of the "full" parameter. I'm not sure all this complication was a step in the right direction...

Do you agree on a major reversion of the related behavior? According to your words, it seems more sensible to make `:droid` work like executing `:give_control` and assigning a side to AI. Actually, I'm also not sure why the behavior of `:droid` is such a complex thing atm...

soliton- commented:
Sorry for the late reply. I was hoping gfgtdf would shed some light on why the current state is the way it is. It makes little sense to me and given ProditorMagnus' comment (which was my understanding as well) and various notes around that imply the opposite of the current controller behavior I'd tend towards aligning behavior with current expectation and documentation. But it would be good to figure out when the behavior was changed and why...

To be clear while the topic is related to the this PR I don't expect change here.

> > Really originally "full" was the only thing. At some point a proxy controller was introduced and then you could not change control fully at all until the introduction of the "full" parameter. I'm not sure all this complication was a step in the right direction...
> 
> Do you agree with a major reversion of the related behavior?
>
> According to your words, it seems more sensible to transfer the functionality of `:droid full` to a new option of `:give_control`, and remove the code that changes the controller status related to OOS within `do_droid`. If I have misunderstood, please share more details about your thoughts on overhauling :droid. Frankly, I am also not sure why the behavior of `:droid` is so complex at the moment...

Reversion would be making `:droid` default to "full" I would think but making `:give_control` do the "full"-part does sound like a good idea.

Mahoyomu committed:
Roughly implemented around :give_control, and removed related funcitionalities in :droid.

There're a lot of redundant comments used for quick locating, and need to be removed in further commits.

Pentarctagon commented:
Any update on this?
